src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -767,155 +775,160 @@ int tempIdentifier(BoundDagEvaluation e)
         /// Make a <see cref="DecisionDag"/> (state machine) starting with the given set of cases in the root node,
         /// and return the node for the root.
         /// </summary>
-        private DecisionDag MakeDecisionDag(ImmutableArray<StateForCase> casesForRootNode)
+        private DecisionDag MakeDecisionDag(ArrayBuilder<StateForCase> casesForRootNode)
         {
             // A work list of DagStates whose successors need to be computed
             var workList = ArrayBuilder<DagState>.GetInstance();
-
-            // A mapping used to make each DagState unique (i.e. to de-dup identical states).
-            var uniqueState = new Dictionary<DagState, DagState>(DagStateEquivalence.Instance);
-
-            // We "intern" the states, so that we only have a single object representing one
-            // semantic state. Because the decision automaton may contain states that have more than one
-            // predecessor, we want to represent each such state as a reference-unique object
-            // so that it is processed only once. This object identity uniqueness will be important later when we
-            // start mutating the DagState nodes to compute successors and BoundDecisionDagNodes
-            // for each one. That is why we have to use an equivalence relation in the dictionary `uniqueState`.
-            DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<BoundDagTemp, IValueSet> remainingValues)
+            try
```

Comment:
Why do we need this `try`? If the issue is there's multiple return points, then we can use a local function instead.

Reply:
we need the try because the worklist was not freed on existing paths.  try/finally was the cleanest and most understandable pattern here for disposing of resources on all codepaths.

We could use a local function, but it jsut kicks the can around.  Now we have to ensure that all releasable data is before the local function, freed after the local function, and that we never insert more code in between in teh future.  So we effectively reimplement C's single-entry/single-exit pattern, and we don't use the construct that solves this naturally which is a core part of C# :)

Reply:
Moved to local function.

Reply:
nit: It looks like there is only one caller of `MakeDecisionDag`. We could also get the instance, pass the `ArrayBuilder` in and free it there.

Reply:
looking.

Reply:
Done.  Thanks!